Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+                // SETTINGS frame whose stream identifier field is anything other than 0x0, the
+                // endpoint MUST respond with a connection error (Section 5.4.1) of type
+                // PROTOCOL_ERROR.
+                throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+            }
+            if flags.contains(.ack) {
+                guard header.length == 0 else {
+                    // When [the ACK flag] is set, the payload of the SETTINGS frame MUST be empty.
+                    // Receipt of a SETTINGS frame with the ACK flag set and a length field value
+                    // other than 0 MUST be treated as a connection error (Section 5.4.1) of type
+                    // FRAME_SIZE_ERROR.
+                    throw NIOHTTP2Errors.ConnectionError(code: .frameSizeError)
+                }
+            } else if header.length == 0 || header.length % 6 != 0 {
+                // A SETTINGS frame with a length other than a multiple of 6 octets MUST be treated
+                // as a connection error (Section 5.4.1) of type FRAME_SIZE_ERROR.
+                throw NIOHTTP2Errors.ConnectionError(code: .frameSizeError)
+            }
+            
+            var settings: [HTTP2Setting] = []
+            settings.reserveCapacity(header.length / 6)
+            
+            var consumed = 0
+            while consumed < header.length {
+                // TODO: name here should be HTTP2SettingsParameter(fromNetwork:), but that's currently defined for NGHTTP2's Int32 value
+                let identifier = HTTP2SettingsParameter(fromPayload: bytes.readInteger()!)
+                let value: UInt32 = bytes.readInteger()!
+                
+                settings.append(HTTP2Setting(parameter: identifier, value: Int(value)))
+                consumed += 6
+            }
+            
+            payload = .settings(settings)
+            
+        case 5:
+            // PUSH_PROMISE frame : RFC 7540 § 6.6
+            guard streamID != .rootStream else {
+                // The stream identifier of a PUSH_PROMISE frame indicates the stream it is associated with.
+                // If the stream identifier field specifies the value 0x0, a recipient MUST respond with a
+                // connection error (Section 5.4.1) of type PROTOCOL_ERROR.
+                throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+            }
+            
+            let stream = HTTP2StreamData.getStream(for: streamID, allocator: self.allocator)
+            var bytesToRead = header.length
+            
+            let padding: UInt8
+            if flags.contains(.padded) {
+                padding = bytes.readInteger()!
+                bytesToRead -= 1
+            } else {
+                padding = 0
+            }
+            
+            let raw: UInt32 = bytes.readInteger()!
+            let promisedStreamID = HTTP2StreamID(knownID: Int32(raw & ~0x8000_0000))
```

Comment:
This streamID is also not allowed to be `.rootStream`.

Reply:
Good catch